gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientWriteChannel.java

```diff
@@ -0,0 +1,206 @@
+            generateWriteOptions(createOptions, channelOptions));
+    writeChannel.setChunkSize(channelOptions.getUploadChunkSize());
+
+    return writeChannel;
+  }
+
+  private class UploadOperation implements Callable<Boolean> {
+
+    // Read end of the pipe.
+    private final InputStream pipeSource;
+    private final StorageResourceId resourceId;
+    private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
+
+    UploadOperation(@Nonnull InputStream pipeSource, @Nonnull StorageResourceId resourceId) {
+      this.resourceId = resourceId;
+      this.pipeSource = pipeSource;
+    }
+
+    @Override
+    public Boolean call() throws Exception {
+      // Try-with-resource will close this end of the pipe so that
+      // the writer at the other end will not hang indefinitely.
+      logger.atFiner().log("Starting upload for resource %s", resourceId);
+      try (pipeSource) {
+        boolean lastChunk = false;
+        ByteBuffer byteBuffer = ByteBuffer.allocate(MAX_BYTES_PER_MESSAGE);
+        while (!lastChunk) {
+          int remainingCapacity = byteBuffer.remaining();
+          ByteString data =
+              ByteString.readFrom(
+                  ByteStreams.limit(pipeSource, remainingCapacity), remainingCapacity);
+          if (data.size() < remainingCapacity) {
+            lastChunk = true;
+          }
+          byteBuffer.put(data.toByteArray());
+          // switch to read mode
+          byteBuffer.flip();
+          // this could result into partial write
+          writeInternal(byteBuffer);
+          if (!lastChunk) {
+            // compact buffer for further writing
+            byteBuffer.compact();
+          }
+        }
+        // last chunk could be partially written
+        // uploading all bytes of last chunk
+        if (lastChunk && byteBuffer.hasRemaining()) {
+          while (byteBuffer.hasRemaining()) {
+            writeInternal(byteBuffer);
+          }
+        }
+        logger.atFiner().log("Uploaded all chunks for resource %s", resourceId);
+        return true;
+      } catch (Exception e) {
+        throw new IOException(
+            String.format("Error occurred while uploading resource %s", resourceId), e);
+      }
+    }
+  }
+
+  private static BlobWriteOption[] generateWriteOptions(
+      CreateObjectOptions createOptions, AsyncWriteChannelOptions channelOptions) {
+    List<BlobWriteOption> writeOptions = new ArrayList<>();
+
+    writeOptions.add(BlobWriteOption.disableGzipContent());
+    writeOptions.add(BlobWriteOption.generationMatch());
+    if (createOptions.getKmsKeyName() != null) {
+      writeOptions.add(BlobWriteOption.kmsKeyName(createOptions.getKmsKeyName()));
+    }
+    if (channelOptions.isGrpcChecksumsEnabled()) {
+      writeOptions.add(BlobWriteOption.crc32cMatch());
+    }
+    return writeOptions.toArray(new BlobWriteOption[writeOptions.size()]);
+  }
+
+  @Override
+  public void close() throws IOException {
```

Comment:
You might want to throw ClosedException if close is called twice?

Reply:
Following what we are doing in `AbstractGoogleAsyncWriteChannel` and returning if not Open.